File: controller-go/internal/controller/scene_controller.go
package controller

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"

	"scenescape-controller/internal/config"
	"scenescape-controller/internal/scene"
	"scenescape-controller/internal/cache"
	"scenescape-controller/pkg/geometry"
	"scenescape-controller/pkg/timestamp"
)

const (
	avgFrames = 100
)

// SceneController is the main controller that orchestrates scene processing
type SceneController struct {
	config     *config.Config
	logger     *logrus.Logger
	ctx        context.Context
	cancel     context.CancelFunc
	wg         sync.WaitGroup
	
	// MQTT client
	mqttClient mqtt.Client
	
	// Cache manager
	cacheManager *cache.Manager
	
	// Scenes and tracking
	scenes        map[string]*scene.Scene
	scenesMutex   sync.RWMutex
	regulateCache map[string]map[string]float64
	
	// Time synchronization
	timeOffset     float64
	lastTimeSync   time.Time
	
	// Tracker configuration
	trackerConfig map[string]interface{}
}

// NewSceneController creates a new scene controller instance
func NewSceneController(cfg *config.Config) (*SceneController, error) {
	if err := cfg.Validate(); err != nil {
		return nil, errors.Wrap(err, "invalid configuration")
	}
	
	ctx, cancel := context.WithCancel(context.Background())
	
	logger := logrus.New()
	if cfg.Verbose {
		logger.SetLevel(logrus.DebugLevel)
	}
	
	sc := &SceneController{
		config:        cfg,
		logger:        logger,
		ctx:           ctx,
		cancel:        cancel,
		scenes:        make(map[string]*scene.Scene),
		regulateCache: make(map[string]map[string]float64),
		trackerConfig: make(map[string]interface{}),
	}
	
	// Load tracker configuration
	if err := sc.loadTrackerConfig(); err != nil {
		return nil, errors.Wrap(err, "failed to load tracker config")
	}
	
	// Initialize cache manager
	cacheManager, err := cache.NewManager(cfg.RestURL, cfg.RestAuth, cfg.RootCert, sc.trackerConfig)
	if err != nil {
		return nil, errors.Wrap(err, "failed to create cache manager")
	}
	sc.cacheManager = cacheManager
	
	// Initialize MQTT client
	if err := sc.initMQTT(); err != nil {
		return nil, errors.Wrap(err, "failed to initialize MQTT")
	}
	
	return sc, nil
}

// Run starts the scene controller and blocks until shutdown
func (sc *SceneController) Run() error {
	sc.logger.Info("Starting SceneScape Controller (Go)")
	
	// Connect to MQTT broker
	if token := sc.mqttClient.Connect(); token.Wait() && token.Error() != nil {
		return errors.Wrap(token.Error(), "failed to connect to MQTT broker")
	}
	
	sc.logger.Infof("Publishing camera visibility info on %s topic", sc.config.VisibilityTopic)
	
	// Start background tasks
	sc.wg.Add(2)
	go sc.timeSync()
	go sc.cacheRefresh()
	
	// Wait for shutdown
	<-sc.ctx.Done()
	return nil
}

// Shutdown gracefully shuts down the controller
func (sc *SceneController) Shutdown() error {
	sc.logger.Info("Shutting down SceneScape Controller")
	
	// Cancel context to stop background tasks
	sc.cancel()
	
	// Disconnect MQTT client
	sc.mqttClient.Disconnect(250)
	
	// Wait for background tasks to complete
	sc.wg.Wait()
	
	sc.logger.Info("Shutdown complete")
	return nil
}

// loadTrackerConfig loads tracker configuration from file
func (sc *SceneController) loadTrackerConfig() error {
	if sc.config.TrackerConfigFile == "" {
		return nil
	}
	
	data, err := os.ReadFile(sc.config.TrackerConfigFile)
	if err != nil {
		return errors.Wrap(err, "failed to read tracker config file")
	}
	
	var trackerCfg config.TrackerConfig
	if err := json.Unmarshal(data, &trackerCfg); err != nil {
		return errors.Wrap(err, "failed to parse tracker config")
	}
	
	// Convert frame-based parameters to time-based
	if trackerCfg.BaselineFrameRate > 0 {
		sc.trackerConfig["max_unreliable_time"] = float64(trackerCfg.MaxUnreliableFrames) / trackerCfg.BaselineFrameRate
		sc.trackerConfig["non_measurement_time_dynamic"] = float64(trackerCfg.NonMeasurementFramesDynamic) / trackerCfg.BaselineFrameRate
		sc.trackerConfig["non_measurement_time_static"] = float64(trackerCfg.NonMeasurementFramesStatic) / trackerCfg.BaselineFrameRate
	}
	
	return nil
}

// initMQTT initializes the MQTT client and sets up callbacks
func (sc *SceneController) initMQTT() error {
	opts := mqtt.NewClientOptions()
	opts.AddBroker(fmt.Sprintf("tcp://%s", sc.config.Broker))
	opts.SetClientID("scenescape-controller-go")
	opts.SetKeepAlive(60 * time.Second)
	opts.SetDefaultPublishHandler(sc.onMessage)
	opts.SetConnectionLostHandler(sc.onConnectionLost)
	opts.SetOnConnectHandler(sc.onConnect)
	
	// Set authentication if provided
	if sc.config.BrokerAuth != "" {
		if err := sc.setMQTTAuth(opts); err != nil {
			return errors.Wrap(err, "failed to set MQTT authentication")
		}
	}
	
	// Set TLS if certificates are provided
	if sc.config.RootCert != "" {
		if err := sc.setMQTTTLS(opts); err != nil {
			return errors.Wrap(err, "failed to set MQTT TLS")
		}
	}
	
	sc.mqttClient = mqtt.NewClient(opts)
	return nil
}

// onConnect is called when MQTT connection is established
func (sc *SceneController) onConnect(client mqtt.Client) {
	sc.logger.Info("Connected to MQTT broker")
	
	// Subscribe to relevant topics
	topics := map[string]byte{
		"scenescape/data/camera/+":     0,
		"scenescape/external/+/+":      0,
		"scenescape/data/sensor/+":     0,
		"scenescape/cmd/scene/update/+": 0,
	}
	
	for topic, qos := range topics {
		if token := client.Subscribe(topic, qos, sc.onMessage); token.Wait() && token.Error() != nil {
			sc.logger.Errorf("Failed to subscribe to %s: %v", topic, token.Error())
		} else {
			sc.logger.Debugf("Subscribed to %s", topic)
		}
	}
	
	// Refresh scenes from cache
	go func() {
		if err := sc.cacheManager.RefreshScenes(); err != nil {
			sc.logger.Errorf("Failed to refresh scenes: %v", err)
		}
	}()
}

// onConnectionLost is called when MQTT connection is lost
func (sc *SceneController) onConnectionLost(client mqtt.Client, err error) {
	sc.logger.Errorf("MQTT connection lost: %v", err)
}

// onMessage handles incoming MQTT messages
func (sc *SceneController) onMessage(client mqtt.Client, msg mqtt.Message) {
	topic := msg.Topic()
	payload := string(msg.Payload())
	
	sc.logger.Debugf("Received message on topic %s", topic)
	
	// Parse JSON payload
	var data map[string]interface{}
	if err := json.Unmarshal([]byte(payload), &data); err != nil {
		sc.logger.Errorf("Failed to parse JSON message: %v", err)
		return
	}
	
	// Route message based on topic
	switch {
	case strings.Contains(topic, "/data/camera/"):
		sc.handleCameraData(topic, data)
	case strings.Contains(topic, "/external/"):
		sc.handleExternalData(topic, data)
	case strings.Contains(topic, "/data/sensor/"):
		sc.handleSensorData(topic, data)
	case strings.Contains(topic, "/cmd/scene/update/"):
		sc.handleSceneUpdate(topic, data)
	default:
		sc.logger.Debugf("Unhandled topic: %s", topic)
	}
}

// handleCameraData processes camera detection data
func (sc *SceneController) handleCameraData(topic string, data map[string]interface{}) {
	// Extract camera ID from topic
	parts := strings.Split(topic, "/")
	if len(parts) < 4 {
		sc.logger.Errorf("Invalid camera topic format: %s", topic)
		return
	}
	cameraID := parts[len(parts)-1]
	
	// Find scene for this camera
	scene := sc.cacheManager.GetSceneByCameraID(cameraID)
	if scene == nil {
		sc.logger.Errorf("No scene found for camera: %s", cameraID)
		return
	}
	
	// Process detections
	if objects, ok := data["objects"].(map[string]interface{}); ok {
		if timestampStr, ok := data["timestamp"].(string); ok {
			timestamp, err := timestamp.ParseISO(timestampStr)
			if err != nil {
				sc.logger.Errorf("Failed to parse timestamp: %v", err)
				return
			}
			
			adjustedTime := sc.adjustTime(timestamp)
			
			for objectType, detections := range objects {
				if detArray, ok := detections.([]interface{}); ok && len(detArray) > 0 {
					scene.ProcessDetections(cameraID, detArray, adjustedTime, objectType)
					
					// Get tracked objects and publish
					trackedObjects := scene.GetTrackedObjects(objectType)
					sc.publishDetections(scene, trackedObjects, adjustedTime, objectType, data, cameraID)
				}
			}
		}
	}
}

// handleExternalData processes external sensor data
func (sc *SceneController) handleExternalData(topic string, data map[string]interface{}) {
	// Extract scene ID and object type from topic
	parts := strings.Split(topic, "/")
	if len(parts) < 4 {
		sc.logger.Errorf("Invalid external topic format: %s", topic)
		return
	}
	
	sceneID := parts[len(parts)-2]
	objectType := parts[len(parts)-1]
	
	scene := sc.cacheManager.GetSceneByUID(sceneID)
	if scene == nil {
		sc.logger.Errorf("No scene found for ID: %s", sceneID)
		return
	}
	
	if timestampStr, ok := data["timestamp"].(string); ok {
		timestamp, err := timestamp.ParseISO(timestampStr)
		if err != nil {
			sc.logger.Errorf("Failed to parse timestamp: %v", err)
			return
		}
		
		adjustedTime := sc.adjustTime(timestamp)
		
		// Process external detection
		detections := []interface{}{data}
		scene.ProcessDetections("external", detections, adjustedTime, objectType)
		
		trackedObjects := scene.GetTrackedObjects(objectType)
		sc.publishDetections(scene, trackedObjects, adjustedTime, objectType, data, "external")
	}
}

// handleSensorData processes sensor data
func (sc *SceneController) handleSensorData(topic string, data map[string]interface{}) {
	// TODO: Implement sensor data handling
	sc.logger.Debugf("Sensor data received on topic %s", topic)
}

// handleSceneUpdate processes scene update commands
func (sc *SceneController) handleSceneUpdate(topic string, data map[string]interface{}) {
	sc.logger.Debug("Scene update command received, refreshing cache")
	go func() {
		if err := sc.cacheManager.RefreshScenes(); err != nil {
			sc.logger.Errorf("Failed to refresh scenes: %v", err)
		}
	}()
}

// publishDetections publishes detection data to various topics
func (sc *SceneController) publishDetections(scene *scene.Scene, objects []scene.MovingObject, timestamp float64, objectType string, data map[string]interface{}, cameraID string) {
	sc.publishSceneDetections(scene, objects, objectType, data)
	sc.publishRegulatedDetections(scene, objects, objectType, data, cameraID)
	sc.publishRegionDetections(scene, objects, objectType, data)
}

// publishSceneDetections publishes to scene-level topics
func (sc *SceneController) publishSceneDetections(scene *scene.Scene, objects []scene.MovingObject, objectType string, data map[string]interface{}) {
	// Build detections list
	pubData := make(map[string]interface{})
	for k, v := range data {
		pubData[k] = v
	}
	
	// TODO: Implement buildDetectionsList equivalent
	pubData["objects"] = objects // Placeholder
	
	objectCount := len(objects)
	cacheID := scene.GetName() + "/" + objectType
	
	// Check if we should publish
	lastCount, exists := scene.GetLastPubCount(cacheID)
	if objectCount > 0 || (exists && lastCount > 0) {
		// Add processing time if available
		if debugStart, ok := data["debug_hmo_start_time"].(float64); ok {
			pubData["debug_hmo_processing_time"] = timestamp.GetEpochTime() - debugStart
		}
		
		topic := fmt.Sprintf("scenescape/data/scene/%s/%s", scene.GetUID(), objectType)
		sc.publishJSON(topic, pubData)
		
		scene.SetLastPubCount(cacheID, objectCount)
	}
}

// publishRegulatedDetections publishes to regulated topics with rate limiting
func (sc *SceneController) publishRegulatedDetections(scene *scene.Scene, objects []scene.MovingObject, objectType string, data map[string]interface{}, cameraID string) {
	if sc.config.VisibilityTopic != "regulated" {
		return
	}
	
	now := timestamp.GetEpochTime()
	maxDelay := 1.0 // 1 second max delay
	
	regulateKey := scene.GetName() + "/" + cameraID + "/" + objectType
	sceneName := scene.GetName()
	
	if sc.regulateCache[sceneName] == nil {
		sc.regulateCache[sceneName] = make(map[string]float64)
	}
	
	lastTime, exists := sc.regulateCache[sceneName][regulateKey]
	if !exists || (now-lastTime) >= maxDelay {
		regulatedData := make(map[string]interface{})
		for k, v := range data {
			regulatedData[k] = v
		}
		
		// TODO: Implement buildDetectionsList equivalent
		regulatedData["objects"] = objects // Placeholder
		
		topic := fmt.Sprintf("scenescape/regulated/scene/%s", scene.GetUID())
		sc.publishJSON(topic, regulatedData)
		
		sc.regulateCache[sceneName][regulateKey] = now
	}
}

// publishRegionDetections publishes region and tripwire events
func (sc *SceneController) publishRegionDetections(scene *scene.Scene, objects []scene.MovingObject, objectType string, data map[string]interface{}) {
	// TODO: Implement region and tripwire event detection
	// This would involve checking objects against regions and tripwires
	sc.logger.Debug("Region detection publishing not yet implemented")
}

// publishJSON publishes a JSON message to the specified topic
func (sc *SceneController) publishJSON(topic string, data interface{}) {
	jsonData, err := json.Marshal(data)
	if err != nil {
		sc.logger.Errorf("Failed to marshal JSON: %v", err)
		return
	}
	
	token := sc.mqttClient.Publish(topic, 0, false, jsonData)
	if token.Wait() && token.Error() != nil {
		sc.logger.Errorf("Failed to publish to %s: %v", topic, token.Error())
	}
}

// adjustTime adjusts timestamp based on configuration
func (sc *SceneController) adjustTime(ts float64) float64 {
	if sc.config.RewriteAllTime {
		return timestamp.GetEpochTime()
	}
	
	adjusted := ts + sc.timeOffset
	now := timestamp.GetEpochTime()
	
	if sc.config.RewriteBadTime && math.Abs(adjusted-now) > sc.config.MaxLag {
		return now
	}
	
	return adjusted
}

// timeSync runs time synchronization in background
func (sc *SceneController) timeSync() {
	defer sc.wg.Done()
	
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()
	
	for {
		select {
		case <-sc.ctx.Done():
			return
		case <-ticker.C:
			// TODO: Implement NTP time synchronization
			// For now, just reset offset to 0
			sc.timeOffset = 0.0
			sc.lastTimeSync = time.Now()
		}
	}
}

// cacheRefresh runs cache refresh in background
func (sc *SceneController) cacheRefresh() {
	defer sc.wg.Done()
	
	ticker := time.NewTicker(60 * time.Second)
	defer ticker.Stop()
	
	for {
		select {
		case <-sc.ctx.Done():
			return
		case <-ticker.C:
			if err := sc.cacheManager.RefreshScenes(); err != nil {
				sc.logger.Errorf("Failed to refresh scenes: %v", err)
			}
		}
	}
}

File: controller-cpp/src/scene_controller.cpp
#include "scene_controller.h"
#include "timestamp.h"
#include "detections_builder.h"
#include <fstream>
#include <chrono>
#include <iostream>

namespace scenescape {

constexpr int AVG_FRAMES = 100;

SceneController::SceneController(
    bool rewrite_bad_time,
    bool rewrite_all_time, 
    double max_lag,
    const std::string& mqtt_broker,
    const std::string& mqtt_auth,
    const std::string& rest_url,
    const std::string& rest_auth,
    const std::string& client_cert,
    const std::string& root_cert,
    const std::string& ntp_server,
    const std::string& tracker_config_file,
    const std::string& schema_file,
    const std::string& visibility_topic
) : rewrite_bad_time_(rewrite_bad_time),
    rewrite_all_time_(rewrite_all_time),
    max_lag_(max_lag),
    broker_(mqtt_broker),
    mqtt_auth_(mqtt_auth),
    ntp_server_(ntp_server),
    visibility_topic_(visibility_topic),
    time_offset_(0.0) {
    
    // Extract tracker configuration
    if (!tracker_config_file.empty()) {
        extractTrackerConfigData(tracker_config_file);
    }
    
    // Initialize MQTT client
    mqtt_client_ = std::make_unique<MqttClient>(
        mqtt_broker, mqtt_auth, client_cert, root_cert
    );
    
    mqtt_client_->setOnConnectCallback([this]() { onConnect(); });
    mqtt_client_->setOnMessageCallback([this](const std::string& topic, const std::string& payload) {
        onMessage(topic, payload);
    });
    
    // Initialize cache manager
    cache_manager_ = std::make_unique<CacheManager>(
        rest_url, rest_auth, root_cert, tracker_config_data_
    );
    
    std::cout << "Publishing camera visibility info on " << visibility_topic_ << " topic.\n";
}

SceneController::~SceneController() = default;

void SceneController::loopForever() {
    mqtt_client_->connect();
    mqtt_client_->loopForever();
}

void SceneController::extractTrackerConfigData(const std::string& tracker_config_file) {
    std::ifstream file(tracker_config_file);
    if (!file.is_open()) {
        throw std::runtime_error("Cannot open tracker config file: " + tracker_config_file);
    }
    
    nlohmann::json config;
    file >> config;
    
    if (config.contains("max_unreliable_frames") && 
        config.contains("baseline_frame_rate") &&
        config.contains("non_measurement_frames_dynamic") &&
        config.contains("non_measurement_frames_static")) {
        
        double baseline_frame_rate = config["baseline_frame_rate"];
        tracker_config_data_["max_unreliable_time"] = 
            config["max_unreliable_frames"].get<double>() / baseline_frame_rate;
        tracker_config_data_["non_measurement_time_dynamic"] = 
            config["non_measurement_frames_dynamic"].get<double>() / baseline_frame_rate;
        tracker_config_data_["non_measurement_time_static"] = 
            config["non_measurement_frames_static"].get<double>() / baseline_frame_rate;
    }
}

void SceneController::onConnect() {
    std::cout << "Connected to MQTT broker\n";
    
    // Subscribe to relevant topics
    mqtt_client_->subscribe("scenescape/data/camera/+");
    mqtt_client_->subscribe("scenescape/external/+/+");
    mqtt_client_->subscribe("scenescape/data/sensor/+");
    mqtt_client_->subscribe("scenescape/cmd/scene/update/+");
    
    // Refresh scenes from cache
    cache_manager_->refreshScenes();
}

void SceneController::onMessage(const std::string& topic, const std::string& payload) {
    try {
        nlohmann::json jdata = nlohmann::json::parse(payload);
        
        // Handle different message types based on topic
        if (topic.find("/data/camera/") != std::string::npos) {
            // Extract camera ID from topic
            size_t pos = topic.find("/data/camera/");
            std::string camera_id = topic.substr(pos + 13);
            
            // Find scene for this camera
            Scene* scene = cache_manager_->getSceneByCameraID(camera_id);
            if (!scene) {
                std::cerr << "No scene found for camera: " << camera_id << std::endl;
                return;
            }
            
            // Process camera detections
            if (jdata.contains("objects") && jdata.contains("timestamp")) {
                double timestamp = getEpochTime(jdata["timestamp"].get<std::string>());
                timestamp = adjustTime(timestamp);
                
                for (const auto& [object_type, detections] : jdata["objects"].items()) {
                    if (detections.is_array() && !detections.empty()) {
                        scene->processDetections(camera_id, detections, timestamp, object_type);
                        
                        // Get tracked objects and publish
                        auto tracked_objects = scene->getTrackedObjects(object_type);
                        publishDetections(scene, tracked_objects, timestamp, object_type, jdata, camera_id);
                    }
                }
            }
        }
        else if (topic.find("/external/") != std::string::npos) {
            // Handle external sensor data
            // Extract scene ID and object type from topic
            size_t pos1 = topic.find("/external/");
            size_t pos2 = topic.find("/", pos1 + 10);
            if (pos2 != std::string::npos) {
                std::string scene_id = topic.substr(pos1 + 10, pos2 - pos1 - 10);
                std::string object_type = topic.substr(pos2 + 1);
                
                Scene* scene = cache_manager_->getSceneByUID(scene_id);
                if (scene && jdata.contains("timestamp")) {
                    double timestamp = getEpochTime(jdata["timestamp"].get<std::string>());
                    timestamp = adjustTime(timestamp);
                    
                    // Process external detection
                    std::vector<nlohmann::json> detections = {jdata};
                    scene->processDetections("external", detections, timestamp, object_type);
                    
                    auto tracked_objects = scene->getTrackedObjects(object_type);
                    publishDetections(scene, tracked_objects, timestamp, object_type, jdata, "external");
                }
            }
        }
        else if (topic.find("/cmd/scene/update/") != std::string::npos) {
            // Scene update command - refresh cache
            cache_manager_->refreshScenes();
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Error processing message: " << e.what() << std::endl;
    }
}

void SceneController::publishDetections(
    Scene* scene, 
    const std::vector<MovingObject>& objects, 
    double timestamp, 
    const std::string& object_type, 
    const nlohmann::json& jdata, 
    const std::string& camera_id
) {
    publishSceneDetections(scene, objects, object_type, jdata);
    publishRegulatedDetections(scene, objects, object_type, jdata, camera_id);
    publishRegionDetections(scene, objects, object_type, jdata);
}

void SceneController::publishSceneDetections(
    Scene* scene, 
    const std::vector<MovingObject>& objects, 
    const std::string& object_type, 
    const nlohmann::json& jdata
) {
    // Build detections list
    nlohmann::json pub_data = jdata;
    pub_data["objects"] = buildDetectionsList(objects, scene, visibility_topic_ == "unregulated");
    
    int object_count = pub_data["objects"].size();
    std::string cache_id = scene->getName() + "/" + object_type;
    
    // Check if we should publish (has objects or last publish had objects)
    if (object_count > 0 || 
        (scene->last_pub_count.count(cache_id) && scene->last_pub_count[cache_id] > 0)) {
        
        // Add processing time if available
        if (jdata.contains("debug_hmo_start_time")) {
            pub_data["debug_hmo_processing_time"] = getEpochTime() - jdata["debug_hmo_start_time"].get<double>();
        }
        
        std::string topic = "scenescape/data/scene/" + scene->getUID() + "/" + object_type;
        mqtt_client_->publish(topic, pub_data.dump());
        
        scene->last_pub_count[cache_id] = object_count;
    }
}

void SceneController::publishRegulatedDetections(
    Scene* scene, 
    const std::vector<MovingObject>& objects, 
    const std::string& object_type, 
    const nlohmann::json& jdata, 
    const std::string& camera_id
) {
    if (visibility_topic_ != "regulated") {
        return;
    }
    
    double now = getEpochTime();
    double max_delay = 1.0; // 1 second max delay for regulated publishing
    
    std::string regulate_key = scene->getName() + "/" + camera_id + "/" + object_type;
    
    if (shouldPublish(
        regulate_cache_[scene->getName()].count(regulate_key) ? 
            std::optional<double>(regulate_cache_[scene->getName()][regulate_key]) : 
            std::nullopt, 
        now, max_delay)) {
        
        nlohmann::json regulated_data = jdata;
        regulated_data["objects"] = buildDetectionsList(objects, scene, true);
        
        std::string topic = "scenescape/regulated/scene/" + scene->getUID();
        mqtt_client_->publish(topic, regulated_data.dump());
        
        regulate_cache_[scene->getName()][regulate_key] = now;
    }
}

void SceneController::publishRegionDetections(
    Scene* scene, 
    const std::vector<MovingObject>& objects, 
    const std::string& object_type, 
    const nlohmann::json& jdata
) {
    // Check for region events and tripwire crossings
    auto tripwire_events = scene->checkTripwireEvents(objects);
    
    for (const auto& event : tripwire_events) {
        nlohmann::json event_data;
        event_data["timestamp"] = getIsoTime();
        event_data["objects"] = nlohmann::json::array({event.object.toJson()});
        event_data["direction"] = event.direction;
        
        std::string topic = "scenescape/event/tripwire/" + scene->getUID() + "/tripwire/" + object_type;
        mqtt_client_->publish(topic, event_data.dump());
    }
    
    // Check region occupancy
    for (const auto& [region_id, region] : scene->getRegions()) {
        int count = 0;
        for (const auto& obj : objects) {
            if (region.contains(obj.getPosition())) {
                count++;
            }
        }
        
        if (count > 0) {
            nlohmann::json region_data = jdata;
            region_data["counts"][object_type] = count;
            region_data["metadata"] = region.toJson();
            
            std::string topic = "scenescape/event/region/" + scene->getUID() + "/" + region_id + "/" + object_type;
            mqtt_client_->publish(topic, region_data.dump());
        }
    }
}

bool SceneController::shouldPublish(
    std::optional<double> last_time, 
    double now, 
    double max_delay
) {
    return !last_time.has_value() || (now - last_time.value()) >= max_delay;
}

void SceneController::syncTime() {
    // TODO: Implement NTP time synchronization
    // For now, use system time
    time_offset_ = 0.0;
    last_time_sync_ = std::chrono::steady_clock::now();
}

double SceneController::adjustTime(double timestamp) {
    if (rewrite_all_time_) {
        return getEpochTime();
    }
    
    double adjusted = timestamp + time_offset_;
    double now = getEpochTime();
    
    if (rewrite_bad_time_ && std::abs(adjusted - now) > max_lag_) {
        return now;
    }
    
    return adjusted;
}

} // namespace scenescape
